Ensure dreps exist prior to delegation

lehins opened the pull request:
# Description

Fixes #4598

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

lehins committed:
Add a new `DelegateeDRepNotRegisteredDELEG` predicate failure

lehins committed:
Optimize DRep unregistration for case when validation is turned off

lehins committed:
Add delegations field to the DRep state

lehins committed:
Change state for `DELEG` rule to `CertState` from `DState`

lehins committed:
Prepare CertState to be amendable by delegation logic

lehins committed:
Change state for `GOVCERT` rule to `CertState` from `VState`

lehins committed:
Implement cleanup of delegations upon DRep unregistration

lehins committed:
Fix couple of tests

lehins committed:
Rename `DelegateeNotRegisteredDELEG` to `DelegateeStakePoolNotRegisteredDELEG`

lehins committed:
Various non-semantic cleanup

lehins committed:
Implement CertState update with DRep delegations for the intra-era hard fork

lehins committed:
Avoid expensive call to `elem` in conformance

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/DRep.hs:
```diff
@@ -125,6 +126,7 @@ data DRepState c = DRepState
   { drepExpiry :: !EpochNo
   , drepAnchor :: !(StrictMaybe (Anchor c))
   , drepDeposit :: !Coin
+  , drepDelegs :: !(Set (Credential 'Staking c))
```
You could add to  cardano-ledger-core CHANGELOG: 
``Add `drepDelegs` to `DRepState` ``

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -207,16 +259,15 @@ conwayDelegTransition = do
+      mCurDelegatee <- checkStakeKeyIsRegistered stakeCred
       checkStakeDelegateeRegistered delegatee
-      pure $ dState {dsUnified = processDelegation stakeCred delegatee dsUnified}
+      pure $ processDelegation stakeCred delegatee $ processUnDelegation stakeCred mCurDelegatee certState
```
Why this undelegation step here, in  Deleg?   Is it in case it's "redelegation", so we cleanup the old one before delegating to the new?

lehins replied:
Correct. We need to make sure that old delegation is removed from the DRep's set with delegations

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs:
```diff
@@ -264,11 +265,14 @@ conwayGovCertTransition = do
+            let paidDeposit = drepState ^. drepDepositL
+            guard (refund /= paidDeposit)
+            pure paidDeposit
+      isJust mDRepState ?! ConwayDRepNotRegistered cred
```
At least once every two PRs I wonder why we don't have `failOnNothing`.  We can discuss when you're back, if there's anything to discuss!

lehins replied:
Probably because it's implementation would be `failOnNothing x = (isJust x ?!)`

However, just for consistency sake with `failOnJust` we could add it.

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -200,13 +231,20 @@ conwayDelegTransition = do
             DelegStakeVote targetPool targetDRep ->
               checkPoolRegistered targetPool >> checkDRepRegistered targetDRep
             DelegVote targetDRep -> checkDRepRegistered targetDRep
+    umElemToDelegatee (UM.UMElem _ _ mPool mDRep) =
```
This could probably even in UMap, seems useful and quite generic.

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -173,18 +182,40 @@ conwayDelegTransition = do
       UM.notMember stakeCred (UM.RewDepUView dsUnified) ?! StakeKeyRegisteredDELEG stakeCred
-    checkStakeKeyIsRegistered stakeCred =
-      UM.member stakeCred (UM.RewDepUView dsUnified) ?! StakeKeyNotRegisteredDELEG stakeCred
+    checkStakeKeyIsRegistered stakeCred = do
```
Now the name became a little bit confusing, because it would seem it's only dealing with whether it's registered or not, but actually returning the Delegatee. But I don't have a better idea and it's inferable from the type anyway.. Probably better than `checkStakeKeyIsRegisteredAndReturnDelegattee` :laughing:

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/DRep.hs:
```diff
@@ -164,6 +169,9 @@ instance Crypto c => FromJSON (DRepState c) where
+      <*> pure mempty
 
 drepExpiryL :: Lens' (DRepState c) EpochNo
 drepExpiryL = lens drepExpiry (\x y -> x {drepExpiry = y})
```
A lens would be nice 
```suggestion
drepExpiryL = lens drepExpiry (\x y -> x {drepExpiry = y})

drepDelegsL :: Lens' (DRepState c) (Set (Credential 'Staking c))
drepDelegsL = lens drepDelegs (\x y -> x {drepDelegs = y})

```

lehins replied:
Added in #4707

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/DRep.hs:
```diff
@@ -8,7 +8,7 @@
+  DRep (DRepCredential, DRepKeyHash, DRepScriptHash, DRepAlwaysAbstain, DRepAlwaysNoConfidence),
   DRepState (..),
   drepExpiryL,
   drepAnchorL,
```
Just in case you had added the lens:

```suggestion
  drepAnchorL,
  drepDelegsL
```

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs:
```diff
@@ -236,57 +240,80 @@ conwayGovCertTransition = do
+        case mDRepState of
+          Nothing -> certState'
+          Just dRepState ->
+            certState'
+              { certDState =
+                  certDState
+                    { dsUnified = drepDelegs dRepState UM.⋪ UM.DRepUView (dsUnified certDState)
+                    }
+              }
```
Maybe slightly easier to follow using lenses?   But I can see this module is not so "lensified" - it doesn't matter, just an idea. 
```suggestion
           certDStateL . dsUnifiedL %~  ((UM.⋪) (drepDelegs dRepState) . UM.DRepUView)
```

lehins committed:
Disable CERTS minitrace conformance tests

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/HardFork.hs:
```diff
@@ -60,8 +65,30 @@ instance
+                    SJust (DRepCredential dRep) ->
+                      let addDelegation _ dRepState =
+                            Just $ dRepState {drepDelegs = Set.insert stakeCred (drepDelegs dRepState)}
+                       in case Map.updateLookupWithKey addDelegation dRep ds of
```
TIL about both `mapAccumWithKey` and `updateLookupWithKey`, very cool!

teodanciu reviewed (approved):
Looks good to me, just a few questions but nothing important - you can safely ignore them.

Lucsanszky committed:
Generate delegs that exist in UMap for dstateSpec

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/GovCert.hs:
```diff
@@ -39,10 +39,11 @@ vStateSpec = constrained' $ \ [var|_dreps|] [var|_commstate|] [var|dormantepochs
-govCertSpec ConwayGovCertEnv {..} vs =
-  let reps = lit $ Map.keysSet $ vsDReps vs
+govCertSpec ConwayGovCertEnv {..} certState =
+  let vs = certVState certState
```
It doesn't seem to make sense to have a `CertState` argument if all we do is use the `VState`

Lucsanszky replied:
I'll leave it as it is for now, @lehins might had his reasons to do this despite what you point out is true.

lehins replied:
Yes, I do have my reasons :smile: 

We'll switch to the same approach for SPOs in the near future, so we'll need access to the `PState` as well.

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/LedgerTypes/Specs.hs:
```diff
@@ -434,7 +441,17 @@ certStateSpec acct epoch = constrained $ \ [var|certState|] ->
     [ satisfies vState (vstateSpec epoch)
     , satisfies pState (pstateSpec epoch)
-    , reify pState psStakePoolParams (\ [var|poolreg|] -> satisfies dState (dstateSpec acct poolreg))
+    , reify
+        pState
+        psStakePoolParams
+        ( \ [var|poolreg|] ->
+            reify
+              vState
+              vsDReps
+              ( \ [var|dreps|] ->
+                  satisfies dState (dstateSpec acct poolreg dreps)
+              )
+        )
```
I'm not sure why this didn't use `match` instead of `reify`. This approach makes it impossible to test a value against this spec (but also that's fine for now, because we only care about generation of values for conformance testing)

Soupstraw replied:
I guess it becomes too long when using `match`, but you can use `sel` instead to pick the field you actually care about

Lucsanszky replied:
I'm not sure either and just followed the pattern established by `reify pState...`. Won't address this now, we can clean up later.

Lucsanszky committed:
Fix `txCertKey` to check for delegatee key as well

Until now we discarded the delegatee cases. This led to the
generation of signals where a delegatee key was already unregistered
in the same signal, resulting in a `DelegateeDRepNotRegisteredDELEG`
or a `DelegateeStakePoolNotRegisteredDELEG` failure.

Lucsanszky committed:
TEMP - Stop generating `UnRegDRep` GovCert cases

This commit can be reverted once the spec is up-to-date with:
https://github.com/IntersectMBO/cardano-ledger/issues/4598

Lucsanszky reviewed (commented):
@teodanciu I marked your remarks as resolved so I can merge. I'll let @lehins to address them.
